Userland/Libraries/LibWeb/WebAssembly/WebAssemblyObject.cpp

```diff
@@ -0,0 +1,306 @@
+        return Wasm::Value { static_cast<double>(number) };
+    }
+    case Wasm::ValueType::F32: {
+        auto number = value.to_double(global_object);
+        if (global_object.vm().exception())
+            return {};
+        return Wasm::Value { static_cast<float>(number) };
+    }
+    case Wasm::ValueType::FunctionReference:
+    case Wasm::ValueType::ExternReference:
+        TODO();
+    }
+
+    VERIFY_NOT_REACHED();
+}
+
+JS::NativeFunction* create_native_function(Wasm::FunctionAddress address, String name, JS::GlobalObject& global_object)
+{
+    // FIXME: Cache these.
+    return JS::NativeFunction::create(
+        global_object,
+        name,
+        [address](JS::VM& vm, JS::GlobalObject& global_object) -> JS::Value {
+            Vector<Wasm::Value> values;
+            Optional<Wasm::FunctionType> type;
+            WebAssemblyObject::s_abstract_machine.store().get(address)->visit([&](const auto& value) { type = value.type(); });
+
+            // Grab as many values as needed and convert them.
+            size_t index = 0;
+            for (auto& type : type.value().parameters()) {
+                auto result = to_webassembly_value(vm.argument(index++), type, global_object);
+                if (result.has_value())
+                    values.append(result.release_value());
+                else
+                    return {};
+            }
+
+            auto result = WebAssemblyObject::s_abstract_machine.invoke(address, move(values));
+            // FIXME: Use the convoluted mapping of errors defined in the spec.
+            if (result.is_trap()) {
+                vm.throw_exception<JS::TypeError>(global_object, "Wasm execution trapped (WIP)");
+                return {};
+            }
+
+            if (result.values().is_empty())
+                return JS::js_undefined();
+
+            if (result.values().size() == 1)
+                return to_js_value(result.values().first(), global_object);
+
+            Vector<JS::Value> result_values;
+            for (auto& entry : result.values())
+                result_values.append(to_js_value(entry, global_object));
+
+            return JS::Array::create_from(global_object, result_values);
+        });
+}
+
+void WebAssemblyModuleInstanceObject::initialize(JS::GlobalObject& global_object)
+{
+    Object::initialize(global_object);
+
+    VERIFY(!m_exports_object);
+    m_exports_object = JS::Object::create_empty(global_object);
+    auto& instance = this->instance();
+    for (auto& export_ : instance.exports()) {
+        export_.value().visit(
+            [&](const Wasm::FunctionAddress& address) {
+                auto function = create_native_function(address, export_.name(), global_object);
+                m_exports_object->define_property(export_.name(), function);
+            },
+            [&](const auto&) {
+                // FIXME: Implement other exports!
+            });
+    }
+
+    define_property("exports", m_exports_object);
```

Comment:
This should be a getter instead (https://webassembly.github.io/spec/js-api/#dom-instance-exports), i.e.:

```cpp
define_native_property("exports", exports_getter, nullptr);
```

and then define a simple getter function that returns `m_exports_object`.